src/librustc_passes/ast_validation.rs

```diff
@@ -78,18 +76,29 @@ struct AstValidator<'a> {
     warning_period_57979_impl_trait_in_proj: bool,
 }
 
+/// With the `new` value in `store`,
+/// runs and returns the `scoped` computation,
+/// resetting the old value of `store` after,
+/// and returning the result of `scoped`.
+fn with<C, T, S>(
+    this: &mut C,
+    new: S,
+    store: impl Fn(&mut C) -> &mut S,
+    scoped: impl FnOnce(&mut C) -> T
+) -> T {
+    let old = mem::replace(store(this), new);
+    let ret = scoped(this);
+    *store(this) = old;
+    ret
+}
```

Comment:
After some thought, could you revert this change?
Yes, it factors out some common functionality, but the price is that it reads like a brain teaser now.

Reply:
Fair :) Will do.

Reply:
Done.